pkg/operator/agent_manager.go

```diff
@@ -1,14 +1,221 @@
+func (k *Kontroller) runDaemonsetUpdate(agentImageRepo string) error {
+	agentDaemonsets, err := k.kc.DaemonSets(k.namespace).List(v1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set(managedByOperatorLabels)).String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(agentDaemonsets.Items) == 0 {
+		// No daemonset, create it
+		runErr := k.createAgentDamonset(agentImageRepo)
+		if runErr != nil {
+			return runErr
+		}
+		// runAgent succeeded, all should be well and converging now
+		return nil
+	}
+
+	// There should only be one daemonset since we use a well-known name and
+	// patch it each time rather than creating new ones.
+	if len(agentDaemonsets.Items) > 1 {
+		glog.Errorf("only expected one daemonset managed by operator; found %+v", agentDaemonsets.Items)
+		return fmt.Errorf("only expected one daemonset managed by operator; found %v", len(agentDaemonsets.Items))
+	}
+
+	agentDS := agentDaemonsets.Items[0]
+
+	var dsSemver semver.Version
+	if dsVersion, ok := agentDS.Annotations[constants.AgentVersion]; ok {
+		ver, err := semver.Parse(dsVersion)
+		if err != nil {
+			return fmt.Errorf("agent daemonset had version annotation, but it was not valid semver: %v[%v] = %v", agentDS.Name, constants.AgentVersion, dsVersion)
+		}
+		dsSemver = ver
+	} else {
+		glog.Errorf("managed daemonset did not have a version annotation: %+v", agentDS)
+		return fmt.Errorf("managed daemonset did not have a version annotation")
+	}
+
+	if dsSemver.LT(version.Semver) {
+		// daemonset is too old, update it
+		// TODO: perform a proper rolling update rather than delete-then-recreate
+		// Right now, daemonset rolling updates aren't upstream and are thus fairly
+		// painful to do correctly. In addition, doing it correctly doesn't add too
+		// much value unless we have corresponding detection/rollback logic.
+		falseVal := false
+		err := k.kc.DaemonSets(k.namespace).Delete(agentDS.Name, &v1.DeleteOptions{
+			OrphanDependents: &falseVal, // Cascading delete
+		})
+		// TODO: this requires delete to be blocking to work. Is it?
+		if err != nil {
+			glog.Errorf("could not delete old daemonset %+v: %v", agentDS, err)
+			return err
+		}
+
+		err = k.createAgentDamonset(agentImageRepo)
+		if err != nil {
+			glog.Errorf("could not create new daemonset: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (k *Kontroller) createAgentDamonset(agentImageRepo string) error {
+	_, err := k.kc.DaemonSets(k.namespace).Create(agentDaemonsetSpec(agentImageRepo))
+	return err
+}
+
+func agentDaemonsetSpec(repo string) *v1beta1.DaemonSet {
+	// Each agent daemonset includes the version of the agent in the selector.
+	// This ensures that the 'orphan adoption' logic doesn't kick in for these
+	// daemonsets.
+	versionedSelector := make(map[string]string)
+	for k, v := range managedByOperatorLabels {
+		versionedSelector[k] = v
+	}
+	versionedSelector[constants.AgentVersion] = version.Version
+
+	return &v1beta1.DaemonSet{
```

Comment:
https://github.com/coreos/container-linux-update-operator/pull/64 should make it more readable.

You can also add `?ts=4` in the url for the same effect.